src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -1892,14 +1890,34 @@ const PRIMITIVES: &[(&str, Res)] = &[
     ("char", Res::PrimTy(hir::PrimTy::Char)),
 ];
 
-fn is_primitive(path_str: &str, ns: Namespace) -> Option<(&'static str, Res)> {
+fn is_primitive(
+    path_str: &str,
+    ns: Namespace,
+    cx: &DocContext<'_>,
+    module_id: DefId,
+) -> Option<(&'static str, Res)> {
     if ns == TypeNS {
-        PRIMITIVES
+        let prim = PRIMITIVES
             .iter()
             .filter(|x| x.0 == path_str)
             .copied()
             .map(|x| if x.0 == "true" || x.0 == "false" { ("bool", x.1) } else { x })
-            .next()
+            .next();
+        prim.or_else(|| {
+            let res = cx
+                .enter_resolver(|resolver| {
+                    resolver.resolve_str_path_error(DUMMY_SP, &path_str, ns, module_id)
+                })
+                .ok()?
+                .1;
+            PRIMITIVES
+                .iter()
+                .find(|(_, prim_res)| match (res, prim_res) {
+                    (Res::PrimTy(ty1), Res::PrimTy(ty2)) if ty1 == *ty2 => true,
+                    (_, _) => false,
+                })
+                .copied()
+        })
```

Comment:
This does not look right. Instead of changing `is_primitive` to go through rustc_resolve, can you change some of the callsites not to use `is_primitive`? The only place that should be doing string comparisons is the ambiguity check for modules vs. primitives.

See also https://github.com/rust-lang/rust/pull/77743#discussion_r502614669.